ledger/sandbox/src/main/scala/com/digitalasset/platform/store/dao/events/TransactionsReader.scala

```diff
@@ -62,41 +62,53 @@ private[dao] final class TransactionsReader(
       filter: FilterRelation,
       verbose: Boolean,
   ): Source[(Offset, GetTransactionsResponse), NotUsed] = {
-    val events: Source[EventsTable.Entry[Event], NotUsed] =
-      PaginatingAsyncStream.streamFrom((startExclusive, Option.empty[Int]), getOffset[Event]) {
-        case (prevOffset, prevNodeIndex) =>
-          val query =
-            EventsTable
-              .preparePagedGetFlatTransactions(sqlFunctions)(
-                startExclusive = prevOffset,
-                endInclusive = endInclusive,
-                filter = filter,
-                pageSize = pageSize,
-                previousEventNodeIndex = prevNodeIndex,
-              )
-              .withFetchSize(Some(pageSize))
-          val rawEventsFuture =
-            dispatcher.executeSql(dbMetrics.getFlatTransactions) { implicit connection =>
-              query.asVectorOf(EventsTable.rawFlatEventParser)
-            }
-          rawEventsFuture.flatMap(
-            rawEvents =>
-              Timed.future(
-                future = Future.traverse(rawEvents)(deserializeEntry(verbose)),
-                timer = dbMetrics.getFlatTransactions.translationTimer,
+
+    val eventsRangeF: Future[EventsRange[Long]] = dispatcher.executeSql(dbMetrics.getRowIdRange) {
+      connection =>
+        EventsRange.rowIdRange(EventsRange(startExclusive, endInclusive))(connection)
+    }
+
+    def streamEvents(eventsRange0: EventsRange[Long]): Source[EventsTable.Entry[Event], NotUsed] =
+      PaginatingAsyncStream.streamFrom(
+        eventsRange0,
+        getEventsRange[Event](eventsRange0.endInclusive)) { eventsRange1 =>
+        val query =
+          EventsTable
+            .preparePagedGetFlatTransactions(sqlFunctions)(
+              range = eventsRange1,
+              filter = filter,
+              pageSize = pageSize,
             )
+            .withFetchSize(Some(pageSize))
+        val rawEvents =
+          dispatcher.executeSql(dbMetrics.getFlatTransactions) { implicit connection =>
+            query.asVectorOf(EventsTable.rawFlatEventParser)
+          }
+        rawEvents.flatMap(
+          es =>
+            Timed.future(
+              future = Future.traverse(es)(deserializeEntry(verbose)),
+              timer = dbMetrics.getTransactionTrees.translationTimer,
           )
+        )
       }
```

Comment:
I would prefer not having nested definitions, unless necessary. It looks like `streamEvents` can be brought out of the function's scope (and made `private`) just by passing in the `filter`, which sounds fine.

Reply:
It looks like you can probably also re-use the function where a flat transaction is used by passing in the proper parser.

Reply:
yes you can, this entire class was not DRY initially, but I can try to make sure we are properly re-using abstractions, it is just the purpose of this PR was NOT cleaning things up.